### PR TITLE
ddl: converts NULL to NOT NULL for column types with NULL data reports err

### LINKED ISSUE #24580: Error message problem that converts NULL to NOT NULL for column types with NULL data
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
create table t(a char(250) default null);
insert into t values();
alter table t modify a char not null;
```

### 2. What did you expect to see? (Required)
`ERROR 1265 (01000): Data truncated for column 'a' at row 1`

### 3. What did you see instead (Required)
`ERROR 1138 (22004): Invalid use of NULL value`

### 4. What is your TiDB version? (Required)

 Release Version: v4.0.0-beta.2-2810-g3e0947fd2
Edition: Community
Git Commit Hash: 3e0947fd21ddb0b83ace1f1d5fdeb48d277846bc
Git Branch: master

### PR DESCRIPTION
**Pull request by Howie59**
### What problem does this PR solve?

Issue Number: close #24580 

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

**Comment by Howie59** [comment]
not ready 4 review

**Comment by xhebox** [comment]
@Howie59 You could `/unhold` once it is prepared for review.
/hold

**Review by xhebox — COMMENTED** [review]
Can we pass `isDataTruncated` to return the correct error? Or even extract the logic of `isDataTruncated` outside. It is weird for me that `checkForNullValue` will rely on  `EnableChangeColumnType` while it does not have to. And the comment for the function is not updated, too.

**Comment by xhebox** [comment]
Have replied on the slack, please check. @Howie59

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -3886,7 +3886,7 @@ func (d *ddl) getModifiableColumnJob(ctx sessionctx.Context, ident ast.Ident, or
 	// We support modifying the type definitions of 'null' to 'not null' now.
 	var modifyColumnTp byte
 	if !mysql.HasNotNullFlag(col.Flag) && mysql.HasNotNullFlag(newCol.Flag) {
-		if err = checkForNullValue(ctx, col.Tp != newCol.Tp, ident.Schema, ident.Name, newCol.Name, col.ColumnInfo); err != nil {
+		if err = checkForNullValue(ctx, true, ident.Schema, ident.Name, newCol.Name, col.ColumnInfo); err != nil {
```
Comment (xhebox) [inline ddl/ddl_api.go:3889]: There is another one check: https://github.com/pingcap/tidb/blob/d2bdfd51b71646dd7c78e437ee40458bef6649cd/ddl/column.go#L1744.

**Review by xhebox — COMMENTED** [review]
Rest LGTM

**Comment by Howie59** [comment]
PTAL @zimulala

File: ddl/db_test.go [hunk 2]
```diff
@@ -6831,3 +6832,23 @@ func (s *testDBSuite8) TestDdlMaxLimitOfIdentifier(c *C) {
...
+
+// Close issue #24580.
+// See https://github.com/pingcap/tidb/issues/24580
+func (s *testDBSuite8) TestIssue24580(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	// Enable column change variable.
+	tk.Se.GetSessionVars().EnableChangeColumnType = true
+	defer func() {
+		tk.Se.GetSessionVars().EnableChangeColumnType = false
```
Comment (wjhuang2016) [inline ddl/db_test.go:6844]: No need to set it back.

File: ddl/db_test.go [hunk 3]
```diff
@@ -6911,3 +6912,20 @@ func (s *testDBSuite8) TestDdlMaxLimitOfIdentifier(c *C) {
...
 
 }
+
+// Close issue #24580.
+// See https://github.com/pingcap/tidb/issues/24580
+func (s *testDBSuite8) TestIssue24580(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	// Enable column change variable.
+	tk.Se.GetSessionVars().EnableChangeColumnType = true
```
Comment (wjhuang2016) [inline ddl/db_test.go:6922]: It's removed now.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**